sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -283,10 +283,13 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
         }
 
         def createCartesianProduct() = {
-          if (joinType.isInstanceOf[InnerLike]) {
-            Some(Seq(joins.CartesianProductExec(planLater(left), planLater(right), condition)))
-          } else {
-            None
+          joinType match {
+            case _: InnerLike =>
+              Some(Seq(joins.CartesianProductExec(planLater(left), planLater(right), condition)))
+            case LeftOuter | RightOuter | FullOuter if condition.isEmpty =>
```

Comment:
Good find! I think we need to correct [CartesianProductExec.joinType](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/CartesianProductExec.scala#L65) as well.

Reply:
It seems we can enhance `EliminateOuterJoin` to make it simpler.